### ISSUE TITLE
[2.0.13, 2.1-rc6] Problem creating null values

### ISSUE THREAD

**Issue by ryan-stateless** [orientechnologies/orientdb#4726]
2.0.13 and 2.1-rc6
I am unable to create null values in my database. For example:

POST http://orientdb:2480/command/database/sql 
```create vertex V content {"some_key": "some_val", "some_null": null}```

HTTP/1.1 200 OK
```{"result":[{"@type":"d","@rid":"#9:3","@version":1,"@class":"V","some_key":"some_val"}]}```

My expectation is that the created record would include the field for "some_null".

Other observations:
Nulls are handled correctly in embedded data:
```create vertex V content {"some_key": "some_val", "some_null": null, "some_embedded_data" : {"some_embedded_key": "some_embedded_value", "some_embedded_null": null}}```

```{"result":[{"@type":"d","@rid":"#9:4","@version":1,"@class":"V","some_key":"some_val","some_embedded_data":{"some_embedded_key":"some_embedded_value","some_embedded_null":null}}]}```

The only way I've been able to actually create a null field is by:
1) In odb studio, go to schema browser
2) Click "+ new record"
3) Click add field
4) Specify a name and string field type
5) Click save
If I query this record I will actually get the null back:

``` select from 9:5```
```{"result":[{"@type":"d","@rid":"#9:5","@version":1,"@class":"V","some_null":null}]}```

However after this, any further attempt, either through studio or the rest api to add other null fields, will fail. ODB will quietly throw away updated fields containing null values.

PUT http://orientdb:2480/document/database/%239%3A5
```{"@type":"d","@rid":"#9:5","@version":1,"@class":"V", "some_null":null, "some_other_null": null}```

HTTP/1.1 200 OK
```{"@type":"d","@rid":"#9:5","@version":2,"@class":"V","some_null":null}```

I have tried fiddling with this to no effect:
ALTER DATABASE custom standardElementConstraints=false

**Comment by matanshukry** [comment]
Are we sure it is what we want?

in Javascript it make sense, since null and undefined are different values;
However in Java, we only have null.

In my opinion, inserting null values should be silently ignored. Running a SELECT query resulting with a column where all values are null is weird.

Also, assuming we do go ahead and allow null as a value, what is the difference between a row with a null value at column A, and a row without column A at all?

comments from the orient team?

**Comment by ryan-stateless** [comment]
A few assorted thoughts:

1) Silently throwing away user data is a terrible design decision. At minimum, the API should return an error indicating an attempt to insert malformed data in the database.

2) Regardless of the design decision, there is still a bug here, as it *is* possible to create null valued fields in some circumstances and not others (as detailed above)

3) Why is there a not_null flag for record properties if it's not intended that properties can be null? The distinction between "mandatory" and "not null" implies that null is intended to be a valid value.

4) Though an embedded blob is a somewhat different beast than a first class record property, it will be inconsistent if nulls are allowed in the former and not the latter.

5) A database product, especially one that advertises schemaless operation, should not limit the users in terms of how they want to represent their data. It should empower them. In my use case the data has some properties with null values, and I need to be able to know that the data I get out of ODB is the same as the data the I put into it.

**Comment by matanshukry** [comment]
A few thoughts back, although we need someone from the orient team to decide on those finally.

1) Usually I agree, but throwing 'null' values is not the same as throwing user data. Also, you can use a flag if you don't want orientdb to 'accept'('silently ignore') those null values (aka NotNull). Again though, design decision.

2) Agreed, but I don't think anyone should create any PR before any decision is made. The current state of both allowing values to be null, and silently ignoring them. is not good. Something needs to change, yes.

3) I understand why you would think that, and you might be right. That said though, currently the not null check against an inserted value of null, while mandatory is checked if there was no value, _and_ there is no default value 

4) I fail to see why; I'm not talking about the content of a blob; if the blob contains 10 bytes, where some of them are null, the value is not (should not be) considered null. I haven't played enough with blobs though. so I may misunderstand you.

5) I do understand, although in my opinion null is not  a value, so getting back some non-existing value doesn't make much sense. Also, consider the current case where you insert 'null' into 2 rows and not 3, although all 3 are gonna be null (what do you except them to be?). Creating another type just to allow null doesn't make sense, you might as well use your own pre-defined value to consider empty (0, "", another column etc..)

In fact, although I do not know every DB out there, I'm not aware of a single DB in which null can be considered having a null, but the value is empty. There are many problems doing that.

**Comment by nagarajasr** [comment]
It is very much possible to create a vertex with null values, just that the particular syntax chosen by @ryan-stateless does not permit it:

```sql
orientdb {db=testdb}> create vertex V content {"some_key": "some_val", "some_null": null}

Created vertex 'V#9:0{some_key:some_val} v1' in 0.104000 sec(s).

orientdb {db=testdb}> create vertex v set some_key="some_val", some_null=null

Created vertex 'V#9:1{some_key:some_val,some_null:null} v1' in 0.013000 sec(s).

orientdb {db=testdb}>

```

Note that storing null values in the DB could lead to wastage of storage space (each field is stored in the record, including field name). IMO it makes sense to have a way to insert data that optimizes storage by not storing null values, while still allowing null values to be inserted via a different command.

note also, it is possible to update a field to null. this does not remove the field (there is a separate syntax for that, too). so if it is possible to somehow end up with a field with null value, it must be possible to create the document with the field value null, too - and this is supported, just not in the particular syntax "create vertex content {....}"

however, there is an inconsistency as pointed out by @ryan-stateless in the way records and embedded records are handled, so perhaps it is a bug - that null values in embedded documents are not stripped out by 'create vertex content {....}' command.

```sql

orientdb {db=testdb}> select @this.toJSON() from V

----+------+--------------------------------------------------------------------------------------------
#   |@CLASS|this
----+------+--------------------------------------------------------------------------------------------
0   |null  |{"@type":"d","@rid":"#9:2","@version":1,"@class":"V","some_key":"some_val"}
1   |null  |{"@type":"d","@rid":"#9:3","@version":1,"@class":"V","some_key":"some_val","some_null":null}
----+------+--------------------------------------------------------------------------------------------

2 item(s) found. Query executed in 0.105 sec(s).

orientdb {db=testdb}> update #9:3 remove some_null

Updated record(s) '1' in 0.029000 sec(s).

orientdb {db=testdb}> select @this.toJSON() from V

----+------+---------------------------------------------------------------------------
#   |@CLASS|this
----+------+---------------------------------------------------------------------------
0   |null  |{"@type":"d","@rid":"#9:2","@version":1,"@class":"V","some_key":"some_val"}
1   |null  |{"@type":"d","@rid":"#9:3","@version":2,"@class":"V","some_key":"some_val"}
----+------+---------------------------------------------------------------------------

2 item(s) found. Query executed in 0.015 sec(s).

```

**Comment by tglman** [comment]
Hi @all, yes that is a bug, we should support null also with `create vertex content { ... }` syntax, PR are welcome

**Comment by nagarajasr** [comment]
@tglman i propose to overload ODocument.mergeMap() method to take another boolean parameter mergeNullFields, which if set to true causes null fields to be merged. then from OrientVertex, we explicitly call the overloaded method with mergeNullFields set to true. this way other code relying on existing behaviour of mergeMap to ignore null values are not affected. WDYT?

**Comment by weinr0ck** [comment]
I don't know if this has already been covered, but this appears to be an issue when using 'set' as well.

**Comment by tglman** [comment]
fixed in 2.1 and develop, will be out with the next hotfix
